core/server/worker/src/main/java/alluxio/worker/block/TieredBlockStore.java

```diff
@@ -691,39 +703,62 @@ private void freeSpaceInternal(long sessionId, long availableBytes, BlockStoreLo
       }
+      // Create move tasks for current tier.
+      List<Callable<Boolean>> moveTasks = new ArrayList<>(toMove.size());
       for (BlockTransferInfo entry : toMove) {
-        long blockId = entry.getBlockId();
-        BlockStoreLocation oldLocation = entry.getSrcLocation();
-        BlockStoreLocation newLocation = entry.getDstLocation();
-        MoveBlockResult moveResult;
-        try {
-          moveResult = moveBlockInternal(Sessions.createInternalSessionId(),
-              blockId, oldLocation, newLocation);
-        } catch (InvalidWorkerStateException e) {
-          // Evictor is not working properly
-          LOG.error("Failed to demote blockId {}, this is temp block", blockId);
-          continue;
-        } catch (BlockAlreadyExistsException e) {
-          continue;
-        } catch (BlockDoesNotExistException e) {
-          LOG.info("Failed to demote blockId {}, it could be already deleted", blockId);
-          continue;
-        }
-        if (moveResult.getSuccess()) {
-          synchronized (mBlockStoreEventListeners) {
-            for (BlockStoreEventListener listener : mBlockStoreEventListeners) {
-              listener.onMoveBlockByWorker(sessionId, blockId, moveResult.getSrcLocation(),
-                  newLocation);
+        moveTasks.add(() -> {
+          long blockId = entry.getBlockId();
+          BlockStoreLocation oldLocation = entry.getSrcLocation();
+          BlockStoreLocation newLocation = entry.getDstLocation();
+
+          try {
+            MoveBlockResult moveResult = moveBlockInternal(Sessions.createInternalSessionId(),
+                blockId, oldLocation, newLocation);
+
+            if (moveResult.getSuccess()) {
+              synchronized (mBlockStoreEventListeners) {
+                for (BlockStoreEventListener listener : mBlockStoreEventListeners) {
+                  listener.onMoveBlockByWorker(sessionId, blockId, moveResult.getSrcLocation(),
+                      newLocation);
+                }
+              }
+              return true;
             }
+          } catch (InvalidWorkerStateException e) {
+            // Evictor is not working properly
+            LOG.error("Failed to demote blockId {}, this is temp block", blockId);
+          } catch (BlockAlreadyExistsException e) {
+            LOG.info("Failed to demote blockId {}, it could be already moved", blockId);
+          } catch (BlockDoesNotExistException e) {
+            LOG.info("Failed to demote blockId {}, it could be already deleted", blockId);
           }
-        }
+          return false;
+        });
+      }
+      // Execute move orders in thread pool.
+      try {
+        List<Future<Boolean>> moveResults = mEvictionExecutor.invokeAll(moveTasks);
+        long failCount = moveTasks.size() - moveResults.stream().filter(
+            (fut) -> {
+              try {
+                return fut.get();
+              } catch (Exception e) {
+                return false;
```

Comment:
Do we need to distinguish between the future failures (which return false) and the future exceptions (which are caught here and will return false)?

Reply:
`invokeAll` guarantees that all returned futures will be completed.
This is merely to satisfy method signature.